Implement --exclude-newer-than

notatallshaw opened the pull request:
Closes https://github.com/pypa/pip/issues/6257
Supplants https://github.com/pypa/pip/pull/12717

Changes from the previous PR:

* `upload-before` -> `exclude-newer-than`
* Uses UTC by default so that is reproducible across environments
* Adds tests

notatallshaw commented:
I've made some changes:

* I've reverted to using local timezone based on @BurntSushi feedback
* I've added help for users on how to explicitly specify a timezone
* I am now passing `--exclude-newer-than` to the isolated build install command

BurntSushi commented on src/pip/_internal/cli/cmdoptions.py:
```diff
@@ -796,6 +797,42 @@ def _handle_dependency_group(
+    This is an optparse.Option callback for the --exclude-newer-than option.
+
+    Parses an ISO 8601 datetime string. If no timezone is specified in the string,
+    UTC timezone is applied automatically for consistency with PyPI upload times.
```
That's why I asked if this was a CLI option or not. :-) Definitely agree there is a difference between specifying something on the CLI versus persisting it in a file to share with others.

What you say is what we do in uv. `--exclude-newer` on the CLI accepts things like `2025-08-06` on the CLI and it is interpreted as a local time in your system time zone. But in `pyproject.toml`, it's required to be an RFC 3339 or RFC 9557 timestamp.

pradyunsg commented on src/pip/_internal/cli/cmdoptions.py:
```diff
@@ -796,6 +797,43 @@ def _handle_dependency_group(
+    """
+    if value is None:
+        return None
+    exclude_newer_than = datetime.datetime.fromisoformat(value)
```
We should have a good error message on invalid values, instead of a traceback.

```
❯ pip install --exclude-newer-than whatever numpy --dry-run
Traceback (most recent call last):
  File "/Users/pradyunsg/Developer/github/pip/.venv/bin/pip", line 8, in <module>
    sys.exit(main())
             ~~~~^^
  File "/Users/pradyunsg/Developer/github/pip/src/pip/_internal/cli/main.py", line 80, in main
    return command.main(cmd_args)
           ~~~~~~~~~~~~^^^^^^^^^^
  File "/Users/pradyunsg/Developer/github/pip/src/pip/_internal/cli/base_command.py", line 159, in main
    return self._main(args)
           ~~~~~~~~~~^^^^^^
  File "/Users/pradyunsg/Developer/github/pip/src/pip/_internal/cli/base_command.py", line 172, in _main
    options, args = self.parse_args(args)
                    ~~~~~~~~~~~~~~~^^^^^^
  File "/Users/pradyunsg/Developer/github/pip/src/pip/_internal/cli/base_command.py", line 154, in parse_args
    return self.parser.parse_args(args)
           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/Users/pradyunsg/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/optparse.py", line 1387, in parse_args
    stop = self._process_args(largs, rargs, values)
  File "/Users/pradyunsg/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/optparse.py", line 1427, in _process_args
    self._process_long_opt(rargs, values)
    ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/pradyunsg/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/optparse.py", line 1501, in _process_long_opt
    option.process(opt, value, values, self)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pradyunsg/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/optparse.py", line 784, in process
    return self.take_action(
           ~~~~~~~~~~~~~~~~^
        self.action, self.dest, opt, value, values, parser)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pradyunsg/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/optparse.py", line 805, in take_action
    self.callback(self, opt, value, parser, *args, **kwargs)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pradyunsg/Developer/github/pip/src/pip/_internal/cli/cmdoptions.py", line 814, in _handle_exclude_newer_than
    exclude_newer_than = datetime.datetime.fromisoformat(value)
ValueError: Invalid isoformat string: 'whatever'
```

notatallshaw replied:
Fixed

ichard26 commented on tests/unit/test_cmdoptions.py:
```diff
@@ -51,3 +57,110 @@ def test_identify_python_interpreter_venv(tmpdir: Path) -> None:
     assert identify_python_interpreter(str(tmpdir / "nonexistent")) is None
+
+
+@pytest.mark.parametrize(
+    "value, expected_check",
+    [
+        # Test with timezone info (should be preserved exactly)
+        (
+            "2023-01-01T00:00:00+00:00",
+            lambda dt: dt
+            == datetime.datetime(2023, 1, 1, 0, 0, 0, tzinfo=datetime.timezone.utc),
+        ),
+        (
+            "2023-01-01T12:00:00-05:00",
+            lambda dt: (
+                dt
+                == datetime.datetime(
+                    *(2023, 1, 1, 12, 0, 0),
+                    tzinfo=datetime.timezone(datetime.timedelta(hours=-5)),
+                )
+            ),
+        ),
+    ],
+)
```
Ugh, I am not a fan of how Black explodes large collection literals. Could we instead only pass the expected datetime object via the decorator and perform the equality check in the test?

notatallshaw replied:
Fixed, as best as I could

ichard26 commented on tests/unit/test_cmdoptions.py:
```diff
@@ -51,3 +57,110 @@ def test_identify_python_interpreter_venv(tmpdir: Path) -> None:
+    assert isinstance(result, datetime.datetime)
+
+    # Check that the date/time components match
+    (
+        expected_year,
+        expected_month,
+        expected_day,
+        expected_hour,
+        expected_minute,
+        expected_second,
+    ) = expected_date_time
```
Ouch, I'd suggest something like this:

```python
@pytest.mark.parametrize(
    "value, expected_date_time",
    [
        # Test basic ISO 8601 formats (timezone-naive, will get local timezone)
        ("2023-01-01T00:00:00", datetime(2023, 1, 1, 0, 0, 0)),
        ("2023-12-31T23:59:59", datetime(2023, 12, 31, 23, 59, 59)),
        # Test date only (will be extended to midnight)
        ("2023-01-01", datetime(2023, 1, 1, 0, 0, 0)),
    ],
)
def test_handle_exclude_newer_than_naive_dates(
    value: str, expected_dt: datetime
) -> None:
    assert result.year == expected_dt.year
    assert result.month == expected_dt.month
    assert result.day == expected_dt.day
    assert result.hour == expected_dt.hour
    assert result.minute == expected_dt.minute
    assert result.second == expected_dt.second
```

or 


```python
@pytest.mark.parametrize(
    "value, expected_date_time",
    [
        # Test basic ISO 8601 formats (timezone-naive, will get local timezone)
        ("2023-01-01T00:00:00", (2023, 1, 1, 0, 0, 0)),
        ("2023-12-31T23:59:59", (2023, 12, 31, 23, 59, 59)),
        # Test date only (will be extended to midnight)
        ("2023-01-01", (2023, 1, 1, 0, 0, 0)),
    ],
)
def test_handle_exclude_newer_than_naive_dates(
    value: str, expected_date_time: tuple[int, int, int, int, int, int]
) -> None:
    assert result.timetuple()[:6] == expected_date_time
```

notatallshaw replied:
Done

ichard26 commented on tests/unit/test_cmdoptions.py:
```diff
@@ -51,3 +57,110 @@ def test_identify_python_interpreter_venv(tmpdir: Path) -> None:
+        ("2023-01-01T00:00:00", (2023, 1, 1, 0, 0, 0)),
+        ("2023-12-31T23:59:59", (2023, 12, 31, 23, 59, 59)),
+        # Test date only (will be extended to midnight)
+        ("2023-01-01", (2023, 1, 1, 0, 0, 0)),
```
Correct me if I'm wrong, but this means if I specify `--exclude-newer-than 2025-08-14` (aka today), then pip will exclude candidates published on August 14, right? That behaves opposite to what I'd expect by the name `--exclude-newer-than <day>` -> exclude anything published _after_ this day.

Should this not be extended to `(2023, 1, 2, 0, 0, 0)` or `(2023, 1, 1, 23, 59, 59)` ...?

ichard26 replied:
> To me your suggestion is the same as saying "not larger than 1" and allowing 1.1 because it is less than 2.

This is personal, but I don't read `--exclude-newer-than` like that. I read it like "ignore any packages published after <date/time>". `exclude` means ignore and `than` is exclusive (as is, it won't ignore... double negatives are hard).

The difference with uv's `--exclude-newer` is that the word "than" is missing, thus it's ambiguous whether it's inclusive or exclusive. I do agree that uv's option name looks too much like a boolean flag.

ichard26 replied:
If my explanation doesn't make sense. I agree with @ofek's opinion on the inclusive/exclusive question: https://github.com/pypa/pip/pull/12717#issuecomment-2159702366

For better or worse, I do think it's important that we match uv in semantics. Otherwise, the two tools will function subtly differently which is bound to become a footgun. I guess that means I am most in favour of `--uploaded-before`. Fun :slightly_smiling_face:

ichard26 replied:
That's fair. That leaves me at renaming the option as `--uploaded-before` and leaving the functionality unchanged. It's semantically clear and maintains the logical simplicity IMO (famous last words). cc @pfmoore

notatallshaw replied:
> The right way, in my view, is to explicitly try to parse as a date object first, before parsing as a datetime. If the user input is a date, treat it as the end of the day, i.e., YYYY-MM-DD 23:59:59. If it's a datetime, assume the user was deliberately being precise, and do what they ask.

In Python if I say `my_datetime < datetime(2025, 8, 14)` then `datetime(2025, 8, 14, 6, 0, 0)` is not a valid datetime, the same is true for SQL `my_datetime < '2025-08-15`.

And the underlying comparison is a datetime, not a date, so doing an implicit type conversion is going to create logical inconsistencies. 

I'm happy to change the name, or I'm willing to disallow passing in a datetime without the time specified. I am strongly -1 on mutating the input or doing implicit type conversion.

notatallshaw replied:
On thinking about it because the field being compared is `upload-time` I do like option names which include "upload" in some way.

Considering all the previous discussion I have two new name suggestions `--uploaded-prior-to` and `--uploaded-earlier-than`, due to the length I prefer `--uploaded-prior-to` and will shortly make a commit which changes it to that.

notatallshaw replied:
I've been thinking hard about this and I am still against this inclusive "date" comparison approach, one problem that demonstrates its complexity is time zones:

1. If I do implicit type conversion and parse the user input as a date and then naively take the date part of the upload-time then this creates wrong values, as the user date and the UTC date will not exactly align unless the user happens to be in UTC
2. If I follow uv's example and add 1 day to the user datetime and then do a comparison with the upload-time that's precarious because, IMO, Python's datetime library design is [not easy to understand](https://discuss.python.org/t/why-is-datetime-utcnow-deprecated/86868/10) where as uv gets to follow jiff, which is a lot more intuitive
3. Which leaves me following @pfmoore's suggestion where I tack on `23:59:59` to the user input (reading the spec I think to be technically correct it would be `23:59:59.999999`) and I can not bring myself to mutate user input in that way, for a few reasons, but the biggest one is having this kind of "smart" comparison (like Python package versions and pre-releases) creates hard to follow logical situations

I am therefore going to use an exclusive comparison, and then this issue of date vs. datetime goes away. Everyone agrees that "uploaded prior to 2025-08-14" is the same as "uploaded prior to 2025-08-14 00:00:00".

Though I appreciate that:

* Everyone else seems to disagree with my viewpoint about inclusive operators
* This breaks semantics with uv's option

If this causes this PR to be rejected that's fine, I won't block another PR if someone wants to follow the inclusive comparison using the semantics everyone else seems to agree on.

pfmoore replied:
To be clear, does this mean you will change the option name to `--uploaded-prior-to`?

zanieb replied:
(If we're going to have different semantics a different name seems better)

ichard26 commented on tests/unit/test_finder.py:
```diff
@@ -574,3 +580,72 @@ def test_find_all_candidates_find_links_and_index(data: TestData) -> None:
+
+    def test_package_finder_create_with_exclude_newer_than(self) -> None:
+        """Test that PackageFinder.create() accepts exclude_newer_than parameter."""
+        session = PipSession()
+        search_scope = SearchScope([], [], no_index=False)
+        link_collector = LinkCollector(session, search_scope)
+        selection_prefs = SelectionPreferences(
+            allow_yanked=False,
+            allow_all_prereleases=False,
+        )
```
I'd prefer if we extended `make_test_finder()` to support `exclude_newer_than` than create a finder manually. These finder unit tests don't seem too useful—although I do appreciate the effort you're putting in to make sure this feature well tested—so let's keep this as simple as it needs to be (as a smoke test).

ichard26 commented on tests/functional/test_exclude_newer.py:
```diff
@@ -0,0 +1,93 @@
+        self, script: PipTestEnvironment, data: TestData
+    ) -> None:
+        """Test that --exclude-newer-than fails with invalid date format."""
+        result = script.pip(
+            "install",
+            "--no-index",
+            "-f",
+            data.packages,
+            "--exclude-newer-than=invalid-date",
+            "simple",
+            expect_error=True,
+        )
```
Pro-tip: you can use `script.pip_install_local()` when you don't want to access the Internet.

```suggestion
        result = script.pip_install_local(
            "--exclude-newer-than=invalid-date", "simple", expect_error=True
        )
```

ichard26 commented on tests/functional/test_exclude_newer.py:
```diff
@@ -0,0 +1,93 @@
+        # Should fail with date parsing error
+        assert "invalid" in result.stderr.lower() or "error" in result.stderr.lower()
+
+    def test_exclude_newer_than_help_text(self, script: PipTestEnvironment) -> None:
+        """Test that --exclude-newer-than appears in help text."""
+        result = script.pip("install", "--help")
+        assert "--exclude-newer-than" in result.stdout
+        assert "datetime" in result.stdout
+
+    @pytest.mark.parametrize("command", ["install", "download", "wheel"])
+    def test_exclude_newer_than_available_in_commands(
+        self, script: PipTestEnvironment, command: str
+    ) -> None:
+        """Test that --exclude-newer-than is available in relevant commands."""
+        result = script.pip(command, "--help")
+        assert "--exclude-newer-than" in result.stdout
```
OK, I appreciate the extensive testing, but I will take issue with these tests. I see the value, but I do not think it is worth the test execution time (remember that functional tests are slow since they need to invoke new pip subprocesses) to double check our CLI help or we've added the option to the right commands. Especially since it's not going to help if we add a new command down the line.

notatallshaw replied:
Makes sense, I purposefully added too many tests and then whittled them down, I missed this one.

ichard26 commented on tests/functional/test_exclude_newer.py:
```diff
@@ -0,0 +1,93 @@
+    ) -> None:
+        """Test exclude-newer functionality against real PyPI with upload times."""
+        # Use a small package with known old versions for testing
+        # requests 2.0.0 was released in 2013
```
Could we pass `--no-deps` or use a package w/o dependencies? (e.g., six) Again, this is nit-picky, but I'd like to keep the tests as quick as reasonably possible. I know our test suite is by design not fast, but every little bit helps.

ichard26 reviewed (changes requested):
This is my first review pass. I'll need to think about the larger design more, but I like where this is going. Thank you for working on this!

notatallshaw commented on src/pip/_internal/models/link.py:
```diff
@@ -300,6 +304,12 @@ def from_json(
         if metadata_info is None:
             metadata_info = file_data.get("dist-info-metadata")
 
+        upload_time: datetime.datetime | None
+        if upload_time_data := file_data.get("upload-time"):
+            upload_time = parse_iso_datetime(upload_time_data)
+        else:
+            upload_time = None
```
This feature doesn't exist in the HTML version of the API, it only exists as addition JSON fields in the spec: https://peps.python.org/pep-0700/#specification

> For the HTML version of the API, there is no change from version 1.0. For the JSON version of the API, the following changes are made:

IMO this was a short sighted choice by the spec authors, as it has prevented certain simple index libraries from supporting this feature. But we can't change the past.

ichard26 replied:
Wow, thanks for the information. /me wonders if it'd be useful to propose a v1.1 of the HTML spec to bring it to feature parity with the JSON API, but I don't have the time for that, haha.

notatallshaw commented:
I believe I have addressed all outstanding review items, I also:

* Added this option to the lock command (was missing because original PR predated lock command)
* Added a section to the user guide
* Added a few more tests

ichard26 commented:
As you already know, I have zero free time at the moment. Realistically, I'm probably not going to be able to review anything on an ongoing and long-term basis. Sorry!

notatallshaw commented on docs/html/user_guide.rst:
```diff
@@ -257,6 +257,61 @@ e.g. http://example.com/constraints.txt, so that your organization can store and
+.. note::
+
+    This option only works with package indexes that provide upload-time metadata
+    (such as PyPI). When upload-time information is not available, packages are not
```
Also this behavior was inherited from this previous PR: https://github.com/pypa/pip/pull/12717.

This has been an open issue for uv: https://github.com/astral-sh/uv/issues/12449

I am fine with explicitly failing if an index doesn't support `upload-time`, but it will cause issues as described in the uv issue.

notatallshaw replied:
Okay, I had a think about this and now it will error out immediately if `--uploaded-prior-to` is provided and `upload-time` is not available on a **remote** index.

But it will not error if you specify a local file or links and provide `--uploaded-prior-to`, for example `pip install . --uploaded-prior-to 2024-01-01`. I think it is important to allow the use case of local packages depending on remote packages and being able to filter those remote packages by upload time.

notatallshaw commented on tests/functional/test_uploaded_prior_to.py:
```diff
@@ -0,0 +1,77 @@
+from tests.lib import PipTestEnvironment, TestData
+
+
+class TestUploadedPriorTo:
```
Added tests for missing `upload-time`.

ichard26 commented:
Mistake?
